packages/repository/src/relations/relation.types.ts

```diff
@@ -68,6 +68,42 @@ export interface HasManyDefinition extends RelationDefinitionBase {
   keyTo?: string;
 }
 
+export interface HasManyThroughDefinition extends RelationDefinitionBase {
+  type: RelationType.hasMany;
+  targetsMany: true;
+
+  /**
+   * The through model of this relation.
+   *
+   * E.g. when a Customer has many Order instances and a Seller has many Order instances,
+   * then Order is through.
+   */
+  through: TypeResolver<Entity, typeof Entity>;
+
+  /**
+   * The foreign key used by the through model to reference the source model.
+   *
+   * E.g. when a Customer has many Order instances and a Seller has many Order instances,
+   * then keyTo is "customerId".
+   * Note that "customerId" is the default FK assumed by the framework, users
+   * can provide a custom FK name by setting "keyTo".
+   */
+  keyTo?: string;
+
+  /**
+   * The foreign key used by the through model to reference the target model.
+   *
+   * E.g. when a Customer has many Order instances and a Seller has many Order instances,
+   * then keyThrough is "sellerId".
+   */
+  keyThrough?: string;
```

Comment:
> I find the names `through`, `keyTo` and `keyThrough` a bit difficult to comprehend.
> (...)
> Have you considered using a nested object to define the "through" parameters, for example:
> (...)

☝️ @codejamninja PTAL, the comment above was up-voted but was not addressed yet.

Reply:
Would it be ok to overload through? If you're not trying to specify the through `keyFrom` or `keyTo` I find it awkward to write . . .

```ts
  through: {
    model: () => Appointment
  }
```

and find the following less awkward

```ts
  through: () => Appointment
```

Reply:
Makes sense. Ideally, I think I'd like to have the following options:

When defining the relation via `@hasMany`, I want the short-cut syntax `through: () => Appointment` but also the full-powered variant `through:  {model: () => Appointment, ... }`.

When the relation metadata is resolved, I want `through: () => Appointment` to be expanded into `{model: () => Appointment}` so that we don't have to deal with the shortcut syntax in all places consuming the metadata.

Perhaps the short syntax can be handled by the decorator, so that it expands `@belongsTo(() => Patient, {through: () => Appointment})` to the longer version?

Thoughts?

Reply:
> I find the names `through`, `keyTo` and `keyThrough` a bit difficult to comprehend.
> 
> Let's take a look at a "simple" relation type where only two models are involved:
> 
> * One side of the relation is described by `source` (an Entity) and `keyFrom` (a property name)
> * The other side of the relation is described by `target` (an Entity) and `keyTo` (a property name)
> 
> In that light, when defining an N-N relation, I would find it more consistent if `keyTo` refered to a property name in the `target` entity (not in the `through` model).
> 
> Have you considered using a nested object to define the "through" parameters, for example:
> 
> ```ts
> {
>   type: RelationType.hasMany, // or hasAndBelongsToMany?
>   targetsMany: true,
> 
>   source: Patient,
>   keyFrom: 'id',
> 
>   target: () => Doctor,
>   keyTo: 'id',
> 
>   through: {
>     model: () => Appointment,
>     keyFrom: 'patientId',
>     keyTo: 'doctorId',
>   }
> }
> ```
> 
> @raymondfeng @strongloop/loopback-next thoughts?

Can this be tabled as a separate pull request?